vision_agent/agent/vision_agent_v2.py

```diff
@@ -241,60 +291,86 @@ def __init__(
         else:
             self.tool_recommender = tool_recommender
         self.verbose = verbose
+        self._working_memory: Dict[str, List[str]] = {}
+        if long_term_memory is not None:
+            if "doc" not in long_term_memory.df.columns:
+                raise ValueError("Long term memory must have a 'doc' column.")
+        self.long_term_memory = long_term_memory
+        self.max_retries = 3
         if self.verbose:
             _LOGGER.setLevel(logging.INFO)
 
     def __call__(
         self,
         input: Union[List[Dict[str, str]], str],
         image: Optional[Union[str, Path]] = None,
+        plan: Optional[List[Dict[str, Any]]] = None,
     ) -> str:
         if isinstance(input, str):
             input = [{"role": "user", "content": input}]
-        code, _ = self.chat_with_tests(input, image)
-        return code
+        results = self.chat_with_workflow(input, image, plan)
+        return results["code"]  # type: ignore
 
-    def chat_with_tests(
+    def chat_with_workflow(
         self,
         chat: List[Dict[str, str]],
         image: Optional[Union[str, Path]] = None,
-    ) -> Tuple[str, str]:
+        plan: Optional[List[Dict[str, Any]]] = None,
```

Comment:
Can you give me an example of this gets used ?

Reply:
I mean the `plan`

Reply:
The plan argument is there if you fail a previous task. So if you ask it, "write program X..." and it gets half way through but fails, you can pass the old plan in (which also contains the old code) as well as feedback in the chat, and it will replan all the unexecuted and failed nodes.

Reply:
Oh got it!